aws_lambda_powertools/utilities/idempotency/persistence/redis.py

```diff
@@ -0,0 +1,440 @@
+        item = {
+            "name": data_record.idempotency_key,
+            "mapping": {
+                self.status_attr: data_record.status,
+                self.expiry_attr: data_record.expiry_timestamp,
+            },
+        }
+
+        if data_record.in_progress_expiry_timestamp is not None:
+            item["mapping"][self.in_progress_expiry_attr] = data_record.in_progress_expiry_timestamp
+
+        if self.payload_validation_enabled:
+            item["mapping"][self.validation_key_attr] = data_record.payload_hash
+
+        now = datetime.datetime.now()
+        try:
+            # |     LOCKED     |         RETRY if status = "INPROGRESS"                |     RETRY
+            # |----------------|-------------------------------------------------------|-------------> .... (time)
+            # |             Lambda                                              Idempotency Record
+            # |             Timeout                                                 Timeout
+            # |       (in_progress_expiry)                                          (expiry)
+
+            # Conditions to successfully save a record:
+
+            # The idempotency key does not exist:
+            #    - first time that this invocation key is used
+            #    - previous invocation with the same key was deleted due to TTL
+            #    - SET see https://redis.io/commands/set/
+
+            logger.debug(f"Putting record on Redis for idempotency key: {item['name']}")
+            encoded_item = self._json_serializer(item["mapping"])
+            ttl = self._get_expiry_second(expiry_timestamp=item["mapping"][self.expiry_attr])
+
+            redis_response = self.client.set(name=item["name"], value=encoded_item, ex=ttl, nx=True)
+
+            # redis_response:True -> Redis set succeed, idempotency key does not exist before
+            # return to idempotency and proceed to handler execution phase. Most cases should return here
+            if redis_response:
+                return
+
+            # redis_response:None -> Existing record on Redis, continue to checking phase
+            # The idempotency key exist:
+            #    - previous invocation with the same key and not expired(active idempotency)
+            #    - previous invocation timed out (Orphan Record)
+            #    - previous invocation record expired but not deleted by Redis (Orphan Record)
+
+            idempotency_record = self._get_record(item["name"])
+
+            # status is completed and expiry_attr timestamp still larger than current timestamp
+            # found a valid completed record
+            if idempotency_record.status == STATUS_CONSTANTS["COMPLETED"] and not idempotency_record.is_expired:
+                raise IdempotencyItemAlreadyExistsError
+
+            # in_progress_expiry_attr exist means status is in_progress, and still larger than current timestamp,
+            # found a vaild in_progress record
```

Comment:
```suggestion
            # If the idempotency record has a status of 'INPROGRESS' and has a valid in_progress_expiry_timestamp
            # (meaning the timestamp is greater than the current timestamp in milliseconds), then we have encountered
            # a valid in-progress record. This indicates that another process is currently handling the request, and
            # to maintain idempotency, we raise an error to prevent concurrent processing of the same request.
```